kernel/syscall.c

```diff
@@ -2080,32 +2080,127 @@ static size_t _count_args(const char* const args[])
     return n;
 }
 
-long myst_syscall_execve(
-    const char* filename,
+long myst_syscall_execveat(
+    int dirfd,
+    const char* pathname,
     char* const argv_in[],
-    char* const envp[])
+    char* const envp[],
+    int flags)
 {
     long ret = 0;
     const char** argv = NULL;
     myst_thread_t* current_thread = myst_thread_self();
+    struct locals
+    {
+        char dirname[PATH_MAX];
+    }* locals = NULL;
     myst_path_t* resolved_path = NULL;
-    const char* resolved_filename = filename;
+    char* resolved_pathname = NULL;
 
-    if (!resolved_filename)
+    if (!pathname || (flags & ~(AT_SYMLINK_NOFOLLOW | AT_EMPTY_PATH)))
         ERAISE(-EINVAL);
 
-    // Resolve relative path
-    if (resolved_filename[0] != '/')
+    if (!(resolved_pathname = (char*)malloc(PATH_MAX)))
```

Comment:
We have a helper function `myst_get_absolute_path_from_dirfd` that resolves `dirfd + pathname` to the real path. No need to repeat in this function.

Reply:
I tried to reuse `myst_get_absolute_path_from_dirfd` but I found it more confusing/repetitive to use that and add additional logic to make it work, rather than start fresh. 

> For example, when the pathname is empty, `myst_get_absolute_path_from_dirfd` would behave differently if flag AT_SYMLINK_NOFOLLOW is set, and won't check if `dirfd` points to a directory, while `execveat` doesn't check the flag at this stage, and needs to check if `dirfd` points to a directory.

Reply:
I think the original implementation of `AT_SYMLINK_NOFOLLOW` is wrong in `myst_get_absolute_path_from_dirfd`. We should fix it instead of working around it. The man page says: 
> If pathname is a symbolic link, do not dereference it: instead operate on the link itself.

This should be consistent across all xxxat syscalls including fchmodat and exeaveat.

Reply:
😢 So I did more reading and still think the current implementation is the most suitable... Here's my reasoning

1. __The original implementation of AT_SYMLINK_NOFOLLOW is wrong in `myst_get_absolute_path_from_dirfd`__

I couldn't figure out what part is wrong, possibly because I don't have enough context on this function (like what's its expected behavior). Could you create a new issue to describe that?

2. __Should use `myst_get_absolute_path_from_dirfd` in execveat__

Please see my new comment

Reply:
Talked to Zijie offline. We found some problems with the current `myst_get_absolute_path_from_dirfd` implementation:

* The name is wrong, the function returns a relative path when dirfd is the current directory
* The checking of `AT_SYMLINK_NOFOLLOW` should be done regardless `pathname` is null or not.
* It should add additional parameter to indicate whether to throw an error if `dirfd + pathname` is a symbolic link and `AT_SYMLINK_NOFOLLOW`  flag is specified. 

I think we should address these problems in this PR rather than creating a clone of the original and later on removing it.